core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java

```diff
@@ -0,0 +1,154 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DataFiles;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Metrics;
+import org.apache.iceberg.PartitionKey;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.Tasks;
+
+public abstract class BaseTaskWriter<T> implements TaskWriter<T> {
+  private final List<DataFile> completedFiles = Lists.newArrayList();
+  private final PartitionSpec spec;
+  private final FileFormat format;
+  private final FileAppenderFactory<T> appenderFactory;
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final long targetFileSize;
+
+  protected BaseTaskWriter(PartitionSpec spec, FileFormat format, FileAppenderFactory<T> appenderFactory,
+                           OutputFileFactory fileFactory, FileIO io, long targetFileSize) {
+    this.spec = spec;
+    this.format = format;
+    this.appenderFactory = appenderFactory;
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.targetFileSize = targetFileSize;
+  }
+
+  @Override
+  public void abort() throws IOException {
+    close();
+
+    // clean up files created by this writer
+    Tasks.foreach(completedFiles)
+        .throwFailureWhenFinished()
+        .noRetry()
+        .run(file -> io.deleteFile(file.path().toString()));
+  }
+
+  @Override
+  public List<DataFile> complete() throws IOException {
+    close();
+
+    if (completedFiles.size() > 0) {
+      return ImmutableList.copyOf(completedFiles);
+    } else {
+      return Collections.emptyList();
+    }
+  }
+
+  protected class RollingFileWriter implements Closeable {
+    private static final int ROWS_DIVISOR = 1000;
+    private final PartitionKey partitionKey;
+
+    private EncryptedOutputFile currentFile = null;
+    private FileAppender<T> currentAppender = null;
+    private long currentRows = 0;
+
+    public RollingFileWriter(PartitionKey partitionKey) {
+      this.partitionKey = partitionKey;
+    }
+
+    public void add(T record) throws IOException {
+      if (currentAppender == null) {
+        openCurrent();
+      }
+
+      this.currentAppender.add(record);
+      this.currentRows++;
+
+      if (shouldRollToNewFile()) {
+        closeCurrent();
+      }
+    }
+
+    private void openCurrent() {
+      if (spec.fields().size() == 0) {
```

Comment:
Unpartitioned writers pass a null partition key. Would it make more sense to use that instead of using `spec`?

Reply:
Yeah,  it makes sense.